addons/udes_stock_refactoring/tests/test_splitting.py

```diff
@@ -0,0 +1,647 @@
+        self.assertEqual(pick2_move_line.result_package_id, cherry_pallet2)
+        self.assertEqual(pick2_move_line.product_id, self.cherry)
+
+        # Check that the original picking has been reused
+        self.assertTrue(self.picking.exists())
+
+    def test_split_pallet_different_products(self):
+        """
+        Reserve self.picking with a pallet containing two different products
+        and check it splits correctly when reserved.
+        """
+        mixed_pallet = self.create_package()
+
+        self.create_quant(
+            self.fig.id, self.test_stock_location_01.id, 5, package_id=mixed_pallet.id
+        )
+        self.create_quant(
+            self.grape.id, self.test_stock_location_01.id, 10, package_id=mixed_pallet.id
+        )
+
+        products_info = [{"product": self.fig, "qty": 5}, {"product": self.grape, "qty": 10}]
+        moves = self.create_move(self.picking, products_info)
+
+        fig_move = moves[0]
+        grape_move = moves[1]
+
+        self.picking.action_assign()
+
+        pick = (fig_move | grape_move).picking_id
+
+        self.assertEqual(len(pick), 1)
+        self.assertEqual(pick.state, "assigned")
+        self.assertEqual(pick.group_id.name, mixed_pallet.name)
+
+        # Check we haven't mangled the moves or move_lines
+        moves = pick.move_lines
+        self.assertEqual(len(moves), 2)
+
+        fig_move = moves.filtered(lambda m: m.product_id == self.fig)
+        self.assertEqual(fig_move.product_uom_qty, 5)
+
+        grape_move = moves.filtered(lambda m: m.product_id == self.grape)
+        self.assertEqual(grape_move.product_uom_qty, 10)
+
+        move_lines = pick.move_line_ids
+        self.assertEqual(len(move_lines), 2)
+
+        fig_moveline = move_lines.filtered(lambda ml: ml.product_id == self.fig)
+        self.assertEqual(fig_moveline.product_qty, 5)
+        self.assertEqual(fig_moveline.package_id, mixed_pallet)
+        self.assertEqual(fig_moveline.result_package_id, mixed_pallet)
+
+        grape_moveline = move_lines.filtered(lambda ml: ml.product_id == self.grape)
+        self.assertEqual(grape_moveline.product_qty, 10)
+        self.assertEqual(grape_moveline.package_id, mixed_pallet)
+        self.assertEqual(grape_moveline.result_package_id, mixed_pallet)
+
+        # Check that the original picking has been reused
+        self.assertTrue(self.picking.exists())
+
+    def test_combine_pickings_at_reserve(self):
+        """
+        Create two pickings for two items on the same pallet. Reserve them
+        simultaneously and check they result in one picking with two moves.
+        """
+        pallet = self.create_package()
+
+        self.create_quant(
+            self.elderberry.id, self.test_stock_location_01.id, 5, package_id=pallet.id
+        )
+        self.create_quant(
+            self.elderberry.id, self.test_stock_location_01.id, 10, package_id=pallet.id
+        )
```

Comment:
You can probably move all the create quants into a setup method, and just reuse them for each test, rather than continually having to repeat them.

Reply:
Yep hadn't done too much with the unit tests since moving them from UDES 11 but will have a go at simplifying/improving them.

Reply:
Have now done this. Apple and Banana pallets and quants moved to setup. Have adjusted most tests to make use of these (so in some tests cherry or elderberry were replaced by apple or banana).